[Tablet Orders] Add unit test coverage for `OrderSearchUICommand`

soundsokay opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: https://github.com/woocommerce/woocommerce-ios/issues/12204

## Description
This PR adds unit tests for `OrderSearchUICommand` which is used for for Order search.


## Steps to reproduce
N/A. The type currently has no unit test coverage. This PR adds coverage by introducing `OrderSearchUICommandTests`.

## Testing information
Run the UnitTests test plan in Xcode. Verify all tests pass.

Check the added coverage by modifying the class under test, `OrderSearchUICommand` and rerunning the tests.

For example, comment out:

`onSelectSearchResult(model, viewController)`

Rerun the tests and and you should see the associated test fail: `testDidSelectSearchResult`.

Try the same thing by commenting out:
`analytics.track(.ordersListSearch, withProperties: ["search": "\(keyword)"])`

Rerun the tests and and you should see the associated test fail: `testSynchronizeModelsTracksAnalytics`


## Screenshots
N/A. This change only involves unit tests.

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

soundsokay committed:
test: add OrderSearchUICommandTests

Add new unit test file and first test.

soundsokay committed:
test: add two more test cases

Added testCreateResultsController and testCreateCellViewModel.

soundsokay committed:
test: add storageManager to initializer and allow injecting the lookUpOrderStatus method for testing

soundsokay committed:
test: Remove extension of OrderSearchUICommand that existed only to inject the lookUpOrderStatus method for testing

Instead initialize the class under test with an instance of MockOrderStatusesStoresManager

soundsokay committed:
test: add sanitize keyword test

soundsokay committed:
test: add test to verify synchronizeModels tracks analytics properly

Add StoresManager and Analytics to the initializer of the class under test to make analytics testable.

soundsokay committed:
test: add testDidSelectSearchResult

joshheald commented on WooCommerce/Classes/ViewRelated/Search/Order/OrderSearchUICommand.swift:
```diff
@@ -1,4 +1,7 @@
 import Yosemite
+import Networking
```
```suggestion
```
We don't need this import, it works fine without it 😊.

In general, things in the WooCommerce target shouldn't import directly from Networking/Storage/Hardware targets, and interface only with Yosemite/WooFoundation.

Yosemite is the business logic bridge to the Networking, Storage, and Hardware targets.[ See the "Separation of Concerns" line in this overview...](https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/architecture-overview.md#design-principles)

By convention, the StorageManagerType below is an exception to this – but I'm afraid I don't know why...

soundsokay replied:
Makes sense. Addressed in commit d744bdb.

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+    private var analyticsProvider: MockAnalyticsProvider!
+    private var analytics: WooAnalytics!
+
+    override func setUpWithError() throws {
```
```suggestion
    override func setUp() {
```

I don't know why Apple gives the throwing variant in the template, but we usually don't need it.

soundsokay replied:
Less is more. Addressed in commit 162fb9a

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+        analytics = WooAnalytics(analyticsProvider: analyticsProvider)
+    }
+
+    override func tearDownWithError() throws {
```
```suggestion
    override func tearDown() {
```

soundsokay replied:
Addressed in commit 162fb9a

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+        super.tearDown()
+    }
+
+    func testCreateStarterViewControllerReturnsNil() throws {
```
```suggestion
    func test_createStarterViewController_returns_nil() {
```
We use snake(ish) case for test names... and I don't think this can throw.

soundsokay replied:
Addressed `throws` in commit 162fb9a

Renamed in commit 1b2dc9f

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+        let starterViewController = command.createStarterViewController()
+
+        // Then
+        XCTAssertNil(starterViewController, "Expected createStarterViewController to return nil")
```
It's not clear _why_ this should be nil. When that's the case, I try to make it clearer in the tests e.g. in this comment, or the test name.
```suggestion
        XCTAssertNil(starterViewController, "Expected createStarterViewController to return nil so that the empty results table will be shown before the search.")
```

soundsokay replied:
Renamed in commit 1b2dc9f. Having not written the original type, I'm not sure about the intent either. My focus has been on verifying the implementation as is.

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+
+    func testCreateStarterViewControllerReturnsNil() throws {
+        // Given
+        let command = OrderSearchUICommand(siteID: siteID, onSelectSearchResult: { _, _ in }, storageManager: storageManager)
```
This is totally up to you... but I often find it's more convenient to use a property or helper function for the most common system under test. 

The benefit is that if the `init` changes we don't need to update loads of test code.

soundsokay replied:
Good call. I DRY'd it out in commit 2af2f41

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+        XCTAssertNil(starterViewController, "Expected createStarterViewController to return nil")
+    }
+
+    func testCreateResultsController() {
```
By convention, we tend to summarise the situation/expectation in the test name too:
```suggestion
    func test_createResultsController_returns_results_controller_with_correct_predicate_and_sort_options() {
```

soundsokay replied:
Renamed in commit 1b2dc9f to follow convention though it kind of doubles up on the string included in the assert.

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+
+    func testCreateCellViewModel() {
+        // Given
+        let mockOrder = MockOrders().makeOrder(status: .onHold, items: [], shippingLines: [], refunds: [], fees: [], taxes: [], customFields: [], giftCards: [])
```
Feel free to use the defaults... they should be generally safe. I've left the differences intact, but perhaps you don't even need to specify the shippingLines here?

```suggestion
        let mockOrder = MockOrders().makeOrder(status: .onHold, shippingLines: [])
```

soundsokay replied:
That's a good change since it increases clarity about the focus of the test.

Addressed in commit 7cd80ee

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+
+    func testSynchronizeModelsTracksAnalytics() throws {
+        // Given
+        let stores = MockStoresManager(sessionManager: .testingInstance)
+        let command = OrderSearchUICommand(siteID: siteID, onSelectSearchResult: { _, _ in }, storageManager: storageManager, analytics: analytics, stores: stores)
```
If you choose to, having this all in properties/setUp makes it a bit easier to read the tests.

soundsokay replied:
Prefer to leave it as it is since it improves clarity about the dependencies needed for this functionality. Unlike the other tests this is the only one that requires mocked analytics.

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+        }
+
+        // Then
+        guard let event = analyticsProvider.receivedEvents.first else {
+            XCTFail("Expected an event but found none")
+            return
+        }
+        XCTAssertEqual(event, "orders_list_search")
+
+        guard let eventProperties = analyticsProvider.receivedProperties.first else {
+            XCTFail("Expected event properties but found none")
+            return
+        }
+        XCTAssertEqual(eventProperties["search"] as? String, keyword)
```
Analytic event testing is so gross in this whole setup... there's nothing wrong with any of what you've done here, just empathising that it's so wordy. We should improve that...

soundsokay replied:
I agree. Would be great to reduce the verbosity.

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+
+    func testDidSelectSearchResult() throws {
+        // Given
+        let order = Networking.Order.fake().copy()
```
```suggestion
        let order = Networking.Order.fake()
```
You only need to `.copy` if you're wanting a different value for one or more properties, e.g. `let order = Networking.Order.fake().copy(status: .processing)`

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+        let orderStatus = viewStorage.insertNewObject(ofType: StorageOrderStatus.self)
+        orderStatus.name = name
+        orderStatus.slug = name
+        orderStatus.siteID = MockOrderStatusesStoresManager.siteID
```
I think it's most reliable if you save these...
```suggestion
        orderStatus.siteID = MockOrderStatusesStoresManager.siteID
        viewStorage.saveIfNeeded()
```

soundsokay replied:
Good tip. Addressed in commit 09efa2e

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Search/Order/OrderSearchUICommandTests.swift:
```diff
@@ -0,0 +1,171 @@
+        storageManager.viewStorage.saveIfNeeded()
+    }
+
+    private func waitFor<T>(timeout: TimeInterval = 1, _ completion: (_ promise: @escaping (T) -> Void) -> Void) -> T? {
+        var result: T?
+        let expectation = self.expectation(description: "Waiting for completion")
+        completion {
+            result = $0
+            expectation.fulfill()
+        }
+        waitForExpectations(timeout: timeout, handler: nil)
+        return result
+    }
```
Why do we need this? There's an [implementation in TestKit that we normally use](https://github.com/woocommerce/woocommerce-ios/blob/445a748833707ba989c465412b421b0a02948e7d/TestKit/Sources/TestKit/XCTestCase%2BWait.swift#L72) – I'm not sure whether you knew about that, or if this offers some different behaviour we want?

Anyway, if you choose to remove it you just need to fix some self-capture errors, and it seems to work correctly otherwise.

soundsokay replied:
I think I needed this because using the referenced implementation was producing errors.

joshheald replied:
What errors do you get? I'm not seeing any. I've pushed up 2d6d131 with the changes to see what happens in CI... if it doesn't work, we can revert.

soundsokay replied:
Thanks for pushing that. Looks good and it's better to reuse the existing code 👍

joshheald reviewed (approved):
Thanks for improving our test coverage!

I've added a bunch of comments – feel free to handle them however you prefer, they're mostly just giving context/conventions that we use since this is only your second PR 😊 

Let me know if anything's unclear

soundsokay committed:
test: remove import and copy()

These are not needed.

soundsokay committed:
test: improve test reliability by saving to mock storage

Per review feedback call saveIfNeeded() on mocked view storage.

soundsokay committed:
test: remove throws

Per code review feedback.

soundsokay committed:
test: improve test names

Per PR feedback, follow convention for verbose test names including expectation details.

soundsokay committed:
test: remove parameters unrelated to the test

soundsokay committed:
test: streamlined the unit test setup by applying the DRY principle

Also renamed to systemUnderTest.

soundsokay commented:
Thanks for your detailed review @joshheald 

All comments have been addressed with code changes except this one:

> There's an [implementation in TestKit that we normally use](https://github.com/woocommerce/woocommerce-ios/blob/445a748833707ba989c465412b421b0a02948e7d/TestKit/Sources/TestKit/XCTestCase%2BWait.swift#L72)

joshheald committed:
Use TestKit `waitFor` function

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/210">19.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>
